File: niko-studio/src/knowledge/services/providers/openai_embedding.py
"""
OpenAI Embedding Provider 实现

使用 OpenAI 官方 SDK，支持 text-embedding-3 系列模型。
"""

import time
from typing import Any

from openai import AsyncOpenAI

from ..models import (
    EmbeddingResponse,
    ProviderType,
    TokenUsage,
    EmbeddingError,
)


# 模型维度映射
MODEL_DIMENSIONS: dict[str, int] = {
    "text-embedding-3-small": 1536,
    "text-embedding-3-large": 3072,
    "text-embedding-ada-002": 1536,
}

# 模型价格 (USD per 1M tokens)
MODEL_PRICING: dict[str, float] = {
    "text-embedding-3-small": 0.02,
    "text-embedding-3-large": 0.13,
    "text-embedding-ada-002": 0.10,
}


class OpenAIEmbeddingProvider:
    """OpenAI Embedding Provider 实现

    支持 text-embedding-3 系列模型，包括自定义维度。
    """

    def __init__(
        self,
        api_key: str | None = None,
        base_url: str | None = None,
        organization: str | None = None,
        default_model: str = "text-embedding-3-small",
        timeout: float = 60.0,
        max_retries: int = 3,
    ) -> None:
        """初始化 OpenAI Embedding Provider"""
        self._client = AsyncOpenAI(
            api_key=api_key,
            base_url=base_url,
            organization=organization,
            timeout=timeout,
            max_retries=max_retries,
        )
        self._default_model = default_model

    @property
    def provider_type(self) -> ProviderType:
        return ProviderType.OPENAI

    def get_dimensions(self, model: str) -> int:
        return MODEL_DIMENSIONS.get(model, 1536)

    async def embed(
        self,
        texts: list[str],
        model: str,
        *,
        dimensions: int | None = None,
    ) -> EmbeddingResponse:
        """执行向量编码请求"""
        kwargs: dict[str, Any] = {
            "model": model or self._default_model,
            "input": texts,
        }

        # text-embedding-3 系列支持自定义维度
        if dimensions and model.startswith("text-embedding-3"):
            kwargs["dimensions"] = dimensions

        start_time = time.perf_counter()
        try:
            response = await self._client.embeddings.create(**kwargs)
        except Exception as e:
            raise EmbeddingError(str(e), provider=ProviderType.OPENAI) from e

        latency_ms = int((time.perf_counter() - start_time) * 1000)

        embeddings = [item.embedding for item in response.data]
        actual_dimensions = len(embeddings[0]) if embeddings else 0

        usage = TokenUsage(
            prompt_tokens=response.usage.prompt_tokens,
            completion_tokens=0,
            total_tokens=response.usage.total_tokens,
            estimated_cost=self._estimate_cost(model, response.usage.total_tokens),
        )

        return EmbeddingResponse(
            embeddings=embeddings,
            model_used=model,
            provider=ProviderType.OPENAI,
            dimensions=actual_dimensions,
            usage=usage,
            latency_ms=latency_ms,
        )

    async def health_check(self) -> bool:
        try:
            await self.embed(texts=["test"], model=self._default_model)
            return True
        except Exception:
            return False

    def _estimate_cost(self, model: str, token_count: int) -> float:
        price = MODEL_PRICING.get(model, 0)
        return token_count * price / 1_000_000

File: src-ts/knowledge/providers/openai-embedding.ts
/**
 * Knowledge module - OpenAI Embedding Provider
 *
 * Implements the EmbeddingProvider interface using the OpenAI API.
 * Supports text-embedding-3 series models with custom dimensions.
 */

import type { EmbeddingResponse } from '../models';
import {
  ProviderType,
  createTokenUsage,
  createEmbeddingResponse,
  EmbeddingError,
} from '../models';

// ============================================================
// Model dimensions mapping
// ============================================================

const MODEL_DIMENSIONS: Record<string, number> = {
  'text-embedding-3-small': 1536,
  'text-embedding-3-large': 3072,
  'text-embedding-ada-002': 1536,
};

// ============================================================
// Model pricing (USD per 1M tokens)
// ============================================================

const MODEL_PRICING: Record<string, number> = {
  'text-embedding-3-small': 0.02,
  'text-embedding-3-large': 0.13,
  'text-embedding-ada-002': 0.10,
};

// ============================================================
// OpenAI Embedding API types
// ============================================================

interface OpenAIEmbeddingResponse {
  data: Array<{ embedding: number[] }>;
  usage: {
    prompt_tokens: number;
    total_tokens: number;
  };
}

// ============================================================
// OpenAI Embedding Provider
// ============================================================

export class OpenAIEmbeddingProvider {
  private readonly _apiKey: string | null;
  private readonly _baseUrl: string | null;
  private readonly _organization: string | null;
  private readonly _defaultModel: string;
  private readonly _timeout: number;
  private readonly _maxRetries: number;

  constructor(params: {
    apiKey?: string | null;
    baseUrl?: string | null;
    organization?: string | null;
    defaultModel?: string;
    timeout?: number;
    maxRetries?: number;
  }) {
    this._apiKey = params.apiKey ?? process.env.OPENAI_API_KEY ?? null;
    this._baseUrl = params.baseUrl ?? null;
    this._organization = params.organization ?? null;
    this._defaultModel = params.defaultModel ?? 'text-embedding-3-small';
    this._timeout = params.timeout ?? 60.0;
    this._maxRetries = params.maxRetries ?? 3;
  }

  get providerType(): string {
    return ProviderType.OPENAI;
  }

  getDimensions(model: string): number {
    return MODEL_DIMENSIONS[model] ?? 1536;
  }

  private _getApiUrl(): string {
    const base = this._baseUrl ?? 'https://api.openai.com/v1';
    return `${base.replace(/\/$/, '')}/embeddings`;
  }

  private _buildHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };
    if (this._apiKey) {
      headers['Authorization'] = `Bearer ${this._apiKey}`;
    }
    if (this._organization) {
      headers['OpenAI-Organization'] = this._organization;
    }
    return headers;
  }

  /**
   * Execute vector encoding request
   */
  async embed(
    texts: string[],
    model: string,
    options?: { dimensions?: number },
  ): Promise<EmbeddingResponse> {
    const body: Record<string, unknown> = {
      model: model || this._defaultModel,
      input: texts,
    };

    // text-embedding-3 series supports custom dimensions
    if (options?.dimensions && model.startsWith('text-embedding-3')) {
      body.dimensions = options.dimensions;
    }

    const startTime = performance.now();
    let response: Response;
    try {
      response = await fetch(this._getApiUrl(), {
        method: 'POST',
        headers: this._buildHeaders(),
        body: JSON.stringify(body),
        signal: AbortSignal.timeout(this._timeout * 1000),
      });
    } catch (e) {
      throw new EmbeddingError((e as Error).message, ProviderType.OPENAI);
    }

    if (!response.ok) {
      const errorText = await response.text();
      throw new EmbeddingError(errorText, ProviderType.OPENAI);
    }

    const data = (await response.json()) as OpenAIEmbeddingResponse;
    const latencyMs = Math.round(performance.now() - startTime);

    const embeddings = data.data.map(item => item.embedding);
    const actualDimensions = embeddings.length > 0 ? embeddings[0].length : 0;

    const usage = createTokenUsage({
      promptTokens: data.usage.prompt_tokens,
      completionTokens: 0,
      totalTokens: data.usage.total_tokens,
      estimatedCost: this._estimateCost(model, data.usage.total_tokens),
    });

    return createEmbeddingResponse({
      embeddings,
      modelUsed: model,
      provider: ProviderType.OPENAI,
      dimensions: actualDimensions,
      usage,
      latencyMs,
    });
  }

  /**
   * Check provider health status
   */
  async healthCheck(): Promise<boolean> {
    try {
      await this.embed(['test'], this._defaultModel);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Estimate request cost
   */
  private _estimateCost(model: string, tokenCount: number): number {
    const price = MODEL_PRICING[model] ?? 0;
    return (tokenCount * price) / 1_000_000;
  }
}
